From 887a99b9bcd547221761cda6f2cd8b9cf528c60d Mon Sep 17 00:00:00 2001
From: Kevin Wolf <kwolf@redhat.com>
Date: Mon, 5 Jul 2010 14:41:02 -0300
Subject: [PATCH 2/9] block: Handle multiwrite errors only when all requests have completed

RH-Author: Kevin Wolf <kwolf@redhat.com>
Message-id: <1278340862-16848-3-git-send-email-kwolf@redhat.com>
Patchwork-id: 10479
O-Subject: [RHEL-6 qemu-kvm PATCH 2/2] block: Handle multiwrite errors only when
	all requests have completed
Bugzilla: 602209
RH-Acked-by: Amit Shah <amit.shah@redhat.com>
RH-Acked-by: Christoph Hellwig <chellwig@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Bugzilla: 602209

Don't try to be clever by freeing all temporary data and calling all callbacks
when the return value (an error) is certain. Doing so has at least two
important problems:

* The temporary data that is freed (qiov, possibly zero buffer) is still used
  by the requests that have not yet completed.
* Calling the callbacks for all requests in the multiwrite means for the caller
  that it may free buffers etc. which are still in use.

Just remember the error value and do the cleanup when all requests have
completed.

Signed-off-by: Kevin Wolf <kwolf@redhat.com>
(cherry picked from commit de189a1b4a471d37a2909e97646654fc9751b52f)
---
 block.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
---
 block.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

diff --git a/block.c b/block.c
index 0e03583..1b6ae4d 100644
--- a/block.c
+++ b/block.c
@@ -1916,14 +1916,11 @@ static void multiwrite_cb(void *opaque, int ret)
 
     if (ret < 0 && !mcb->error) {
         mcb->error = ret;
-        multiwrite_user_cb(mcb);
     }
 
     mcb->num_requests--;
     if (mcb->num_requests == 0) {
-        if (mcb->error == 0) {
-            multiwrite_user_cb(mcb);
-        }
+        multiwrite_user_cb(mcb);
         qemu_free(mcb);
     }
 }
-- 
1.7.0.3

